wasm/src/main/java/com/dylibso/chicory/wasm/io/InputStreamWasmInputStream.java

```diff
@@ -0,0 +1,139 @@
+package com.dylibso.chicory.wasm.io;
+
+import java.io.BufferedInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+
+/**
+ * A WASM input stream based on {@link InputStream}.
+ */
+class InputStreamWasmInputStream extends WasmInputStream {
+
+    private long position;
+    private final BufferedInputStream is;
+    private final ByteBuffer littleBuffer =
+            ByteBuffer.wrap(new byte[8]).order(ByteOrder.LITTLE_ENDIAN);
+
+    InputStreamWasmInputStream(final InputStream is) {
+        this.is =
+                is instanceof BufferedInputStream
+                        ? (BufferedInputStream) is
+                        : new BufferedInputStream(is);
+    }
+
+    public long position() {
+        return position;
+    }
+
+    public void skip(long count) {
+        // todo: java 12+: is.skipNBytes(count);
+        if (count <= 0) {
+            return;
+        }
+
+        try {
+            long t;
+            while (count > 0) {
+                t = is.skip(count);
+                if (t == 0) {
+                    if (is.read() == -1) {
+                        throw new WasmEOFException();
+                    }
+                    t = 1;
+                }
+                count -= t;
+                position += t;
+            }
+        } catch (IOException e) {
+            throw new WasmIOException(e);
+        }
+    }
+
+    public int rawByteOpt() {
+        try {
+            int res = is.read();
+            if (res >= 0) {
+                position++;
+                return res & 0xff;
+            } else {
+                return -1;
+            }
+        } catch (IOException e) {
+            throw new WasmIOException(e);
+        }
+    }
+
+    public int peekRawByteOpt() {
+        try {
+            is.mark(1);
+            int res = is.read();
+            is.reset();
+            return res;
+        } catch (IOException e) {
+            throw new WasmIOException(e);
+        }
+    }
+
+    private static final byte[] NO_BYTES = new byte[0];
+
+    public byte[] rawBytes(final int len) {
+        if (len == 0) {
+            return NO_BYTES;
+        }
+        // todo: java 12+: return is.readNBytes(len);
+        return readAll(new byte[len], 0, len);
+    }
+
+    public void rawBytes(final ByteBuffer buf) {
+        if (buf.hasArray()) {
```

Comment:
I see around a bunch of optimizations on the usage of the `ByteBuffer`, would it make sense to refactor those in a separate utility class so that they apply to any usage of `ByteBuffer` in the codebase?
Looks a bit of complexity and it should probably be tested thoroughly to ease any debugging in the future.

Reply:
I don't know of any other place or way to factor this out, nor any other use site where this would be used since it's specific to this class. Did you have something in mind? What optimizations are you referring to?

Reply:
Oh, yes, I see. The slow path is *very* slow though, so we'd want to avoid it if possible.

I don't think there's a direct equivalent in `Memory` but we could always introduce a utility class in a later follow-up if needed.

Reply:
Let's keep this in mind in the future, thanks for going through it 👍